### PR TITLE
Detect and warn about crafting with rotten food

### PR DESCRIPTION
**Pull request by jbytheway**
#### Summary
SUMMARY: Interface "Detect and warn about crafting with rotten food"

#### Purpose of change
Partial solution to #33385.

To deal with the long-standing issue of accidentally using rotting food in crafting (especially when crafting a large batch) and thus wasting time and materials.

#### Describe the solution
Based on [idea](https://github.com/CleverRaven/Cataclysm-DDA/pull/36263#pullrequestreview-335708128) of @ifreund.

Allow a recipe filter to exclude rotten components.

Use this new feature when selecting items for crafting to prefer non-rotten ones.  When non-rotten ones do not suffice, warn the player of this fact before proceeding.

In the crafting GUI, mark crafts in brown when they can only be crafted with rotten ingredients.

#### Describe alternatives you've considered
The brown text might be difficult to read; if so, suggested alternatives are welcome.

This only catches the case where the ingredients are rotten when you start the craft.  There is also the situation where things start non-rotten, but become rotten while you craft.  This change does not attempt to handle that, but it could now be handled by a localized change in `recipe::get_component_filter`.

It would be nice for the specific rotten ingredients to be highlighted in the crafting GUI, but that would require fairly invasive changes elsewhere, and certainly shouldn't be part of this PR.  One issue that arises with the current implementation is that the GUI might suggest you have (say) two options to choose from, but if one is rotten then it will pick the other automatically without asking you, which might be surprising.

It would be nice to have unit tests for this, but the player interactivity aspects make that difficult.

#### Testing
Ran unit tests.

In-game experiments with milk crafting.

#### Additional context
See how crafts are listed in brown when they require rotten ingredients: 
rotten-crafting-gui-normal
And when crafting beyond a particular number of batches requires rotten ingredients the higher numbers are brown similarly:
rotten-crafting-gui-batch
If you go ahead and craft, you get this warning:
rotten-craft-warning
And there's a similar warning for continuing an interrupted craft:
rotten-continue-warning

### CODE REVIEW

**Comment by Morat2255** [comment]
Does this apply to things like Lamp Oil that intentionally allow for rotten ingredients for a non-rotten product? That could get annoying, and in some cases confusing.

**Comment by Lamandus** [comment]
> 
> 
> Does this apply to things like Lamp Oil that intentionally allow for rotten ingredients for a non-rotten product? That could get annoying, and in some cases confusing.

there I would like to see the opposite! I want to use the rotten stuff, so I don't loose the "fresh" one, which I would use for food cooking.

**Comment by 1n17** [comment]
> there I would like to see the opposite! I want to use the rotten stuff, so I don't loose the "fresh" one, which I would use for food cooking.
> Something like "prefer rotten components"
+1
Probably could be done in crafting logic, something like - if craft allows rotten and result is food prioritize not rotten / warn and if its not food prefer rotten.

**Comment by jbytheway** [comment]
> Does this apply to things like Lamp Oil that intentionally allow for rotten ingredients for a non-rotten product? That could get annoying, and in some cases confusing.

Yes, this is handled correctly, in that it doesn't warn you about it.  There's no prioritization of ingredient choice based on age for any crafting.  That might indeed be a useful thing, but I don't want to tackle it in this PR.

rotten-fat
rotten-lamp-oil

File: src/recipe.cpp [hunk 1]
```diff
@@ -500,14 +500,19 @@ bool recipe::will_be_blacklisted() const
...
 
-std::function<bool( const item & )> recipe::get_component_filter() const
+std::function<bool( const item & )> recipe::get_component_filter(
+    const recipe_filter_flags flags ) const
 {
     const item result = create_result();
 
     // Disallow crafting of non-perishables with rotten components
     // Make an exception for items with the ALLOW_ROTTEN flag such as seeds
+    bool recipe_forbids_rotten =
```
Comment (ifreund) [inline src/recipe.cpp]:
```suggestion
    const bool recipe_forbids_rotten =
```

File: src/recipe.cpp [hunk 2]
```diff
@@ -500,14 +500,19 @@ bool recipe::will_be_blacklisted() const
...
+        result.is_food() && !result.goes_bad() && !has_flag( "ALLOW_ROTTEN" );
+    bool flags_forbid_rotten =
```
Comment (ifreund) [inline src/recipe.cpp]:
```suggestion
    const bool flags_forbid_rotten =
```

**Comment by snipercup** [comment]
Oops, I actually referenced this PR instead of #36611 about 34 times

**Pull request merged**